### PR TITLE
Add full change check for rabbitmq policy

### LINKED ISSUE #1566: Ansible barfs if ANSIBLE_HOSTS file is +x
If your ANSIBLE_HOSTS file ends up being +x for some reason, it appears that ansible... attempts to run it as a program?

### LINKED ISSUE #29264: rabbitmq_policy doesn't update a policy that already exists
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
rabbitmq_policy

##### ANSIBLE VERSION
`2.3`

##### SUMMARY
From @tedgin on 2016-01-28T18:29:35Z
Assume we have a policy named `user_queues` that applies to all queues matching the pattern `^prefix_`.  Requirements change and `user_queues` needs to be applied to queues matching the new pattern `/`.  If the play using `rabbitmq_policy` to set policy is modified to set the new pattern and run, it will run successfully, but the `user_queues` policy remains unchanged.

It would be better if `rabbitmq_policy` would change policies that already exists or at least generate a warning or error.

Copied from original issue: ansible/ansible-modules-extras#1566

### PR DESCRIPTION
**Pull request by puzan**
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

rabbitmq_policy
##### ANSIBLE VERSION

```
ansible 2.1.0 (devel 7e3963420d) last updated 2016/05/15 11:11:28 (GMT +300)
  lib/ansible/modules/core: (detached HEAD 25414bbd1b) last updated 2016/05/15 12:07:46 (GMT +300)
  lib/ansible/modules/extras: (detached HEAD e13942f9c5) last updated 2016/05/15 12:07:46 (GMT +300)
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

Fixes #29264.

Add check for all policy fields (name, apply_to, pattern, tags, priority) to have correct changed state. Previosly changed state was based on policy name only.

Test playbook (pattern change):

``` yaml
- name: Update policy
  hosts: localhost
  tasks:
    - name: add policy
      rabbitmq_policy: >
        name=HA
        pattern='.*'
        tags="ha-mode=all"
        node=rabbit@localhost
    - name: update policy
      rabbitmq_policy: >
        name=HA
        pattern='test'
        tags="ha-mode=all"
        node=rabbit@localhost
```

Before fix:

```
PLAY [Update policy] ***********************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [add policy] **************************************************************
changed: [localhost]

TASK [update policy] ***********************************************************
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0
```

After fix:

```
PLAY [Update policy] ***********************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [add policy] **************************************************************
changed: [localhost]

TASK [update policy] ***********************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=2    unreachable=0    failed=0
```

### CODE REVIEW

**Comment by puzan** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/2236

**Comment by puzan** [comment]
It fixes https://github.com/ansible/ansible-modules-extras/issues/1566

**Comment by puzan** [comment]
Hello @retr0h, @bcoca, @jimi-c!

Who can review this old PR?

File: lib/ansible/modules/messaging/rabbitmq_policy.py [hunk 1]
```diff
@@ -51,13 +51,15 @@
     version_added: "2.1"
   pattern:
     description:
-      - A regex of queues to apply the policy to.
-    required: true
+      - A regex of queues to apply the policy to. Required when
+        C(state=present).
+    required: false
```
Comment (bcoca) [inline lib/ansible/modules/messaging/rabbitmq_policy.py]: this is changing the required/default from how the module used to work, you need to update the description and note that this changes in 2.4 vs the previous 'required' behaviour

**Review by bcoca — CHANGES_REQUESTED** [review]

**Comment by bcoca** [comment]
cc @chrishoffman @manuel-sousa @romanek-adam for comments/review

File: lib/ansible/modules/messaging/rabbitmq_policy.py (hunk 1 above)
Comment (puzan) [reply lib/ansible/modules/messaging/rabbitmq_policy.py]: @bcoca I have updated description. Is there 'version_changed' tag like 'version_added'?

**Comment by puzan** [comment]
ready_for_review

**Comment by puzan** [comment]
bot_status

**Comment by puzan** [comment]
ready_for_review

**Comment by puzan** [comment]
ready_for_review

File: lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py [hunk 2]
```diff
@@ -93,25 +101,55 @@ def __init__(self, module, name):
         self._node = module.params['node']
         self._rabbitmqctl = module.get_bin_path('rabbitmqctl', True)
 
-    def _exec(self, args, run_in_check_mode=False):
+        self._version = self._rabbit_version()
+
+    def _exec(self, args, run_in_check_mode=False, split_lines=True):
         if not self._module.check_mode or (self._module.check_mode and run_in_check_mode):
             cmd = [self._rabbitmqctl, '-q', '-n', self._node]
             args.insert(1, '-p')
```
Comment (barryib) [inline lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py]: @puzan `rabbitmqctl status` doesn't support `vhost`. I run RabbitMQ 3.6.16.

As quick fix, I added `add_vhost=True` as default argument to `_exec` to have something like this :

```python
def _exec(self, args, run_in_check_mode=False, split_lines=True, add_vhost=True):
    .... some code here ....
    if add_vhost:
        args.insert(1, '-p')
        args.insert(2, self._vhost)
```

**Review by barryib — CHANGES_REQUESTED** [review]

**Comment by samdoran** [comment]
!waffling needs_revision

**Comment by barryib** [comment]
@puzan @samdoran It would be nice to do some kind of sanitization for policy tags. In fact, for ansible, almost everything is a string. So it is difficult to use variable to define integer tags.

```yml
# Variable
my_tags:
  expires: 604800000
  ha-mode: exactly
  ha-params: "{{ (( rab_nodes_list | length ) / 2) | int + 1 }}"
  ha-sync-mode: manual

# Policy definition
rabbitmq_policy:
  name: ha
  pattern: 'queue.*'
  priority: 10
  apply_to: queues
  state: present
  tags: "{{ my_tags }}"
```

`ha-params` must be an integer. But with ansible, `{{ (( rab_nodes_list | length ) / 2) | int + 1 }}` return a string.

File: lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py (hunk 2 above)
Comment (puzan) [reply lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py]: I have added your solution.

**Comment by barryib** [comment]
@puzan great.

@puzan @samdoran @bcoca What about my second suggestion for sanitization ? Should I open a different PR for that ?

Actually, here is how I fixed the issue with integer in rabbitmq policy 👍 

```python
# Sanitize function
def _sanitize_tags(self, tags):
    sanitezed_tags = tags
    for key, value in sanitezed_tags.items():
        if isinstance(value, str) and value.isdigit():
            sanitezed_tags[key] = int(value)
    return sanitezed_tags

# Module init
def __init__(self, module, name):
    ....
    self._tags = self._sanitize_tags(module.params['tags'])
    ....
```

BTW, any chance to see this PR merged ? It's almost 3 years old.

**Comment by samdoran** [comment]
@barrylib The best way to validate a field like that is to create a custom validation function. You can use your existing function, just change the argument_spec of tags.

```python
...
    tags={'type': _sanitize_tags, 'required': True}
...
```

This should be in a separate PR.

**Comment by samdoran** [comment]
@barryib Since this is a community module, we rely on votes from community members to gauge if it's appropriate to merge. Please [add your vote](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#when-will-your-pull-request-be-merged).

File: lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py [hunk 3]
```diff
@@ -38,12 +38,17 @@
     version_added: "2.1"
   pattern:
     description:
-      - A regex of queues to apply the policy to.
-    required: true
+      - A regex of queues to apply the policy to. Required when
+        C(state=present). 'Required' flag was changed to 'false' in 2.8.
```
Comment (samdoran) [inline lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py]:
```suggestion
        C(state=present). This option is no longer required as of Ansible 2.9.
```

File: lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py [hunk 4]
```diff
@@ -38,12 +38,17 @@
...
+        version.
+    required: false
+    default: null
   tags:
     description:
-      - A dict or string describing the policy.
-    required: true
+      - A dict or string describing the policy. Required when
+        C(state=present). 'Required' flag was changed to 'false' in 2.8.
```
Comment (samdoran) [inline lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py]:
```suggestion
        C(state=present). This option is no longer required as of Ansible 2.9.
```

**Review by samdoran — CHANGES_REQUESTED** [review]
Please create a changelog fragment and we can merge this. See [this fragment](https://github.com/ansible/ansible/blob/stable-2.7/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

File: lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py (hunk 3 above)
Comment (puzan) [reply lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py]: Fixed

File: lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py (hunk 4 above)
Comment (puzan) [reply lib/ansible/modules/messaging/rabbitmq/rabbitmq_policy.py]: Fixed

**Comment by puzan** [comment]
@samdoran I have added changelog and apply your comments.

**Comment by puzan** [comment]
ready_for_review

**Pull request merged**